File: src/main/java/org/myfintech/payment/entity/Payment.java
/*
 * Copyright (c) 2024 MyFintech Payment Service
 * All rights reserved.
 *
 * This software is proprietary and confidential. Unauthorized copying of this file,
 * via any medium, is strictly prohibited.
 * @author : Dhanuka Ranasinghe
 * @since : Date: 11/07/2025
 */
package org.myfintech.payment.entity;

import static org.myfintech.payment.entity.EntityColumnConstants.COMMON_VERSION;
import static org.myfintech.payment.entity.EntityColumnConstants.FOREIGN_KEY_CONTRACT_ID;
import static org.myfintech.payment.entity.EntityColumnConstants.FOREIGN_KEY_PAYMENT_TRACKING_ID;
import static org.myfintech.payment.entity.EntityColumnConstants.TABLE_PAYMENT;

import java.math.BigDecimal;
import java.time.LocalDate;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.Size;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@EqualsAndHashCode(onlyExplicitlyIncluded = true,callSuper = true)
@Entity
@Table(name = TABLE_PAYMENT)
public class Payment extends AbstractEntity<Long> {

    @Version
    @Column(name = COMMON_VERSION, nullable = false)
    private Integer version;

    @NotNull private LocalDate paymentDate;

    @NotNull
    @Positive
    @Column(precision = 10, scale = 2)
    private BigDecimal amount;

    @NotBlank
    @Size(max = 50)
    @EqualsAndHashCode.Include
    private String type;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = FOREIGN_KEY_CONTRACT_ID, insertable = false, updatable = false)
    private Contract contract;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = FOREIGN_KEY_PAYMENT_TRACKING_ID, insertable = false, updatable = false)
    private PaymentTracking paymentTracking;

    //efficient for retrieval
    @Column(name = FOREIGN_KEY_CONTRACT_ID, nullable = false)
    @EqualsAndHashCode.Include
    private Long contractId;

    @Column(name = FOREIGN_KEY_PAYMENT_TRACKING_ID, nullable = false)
    @EqualsAndHashCode.Include
    private Long trackingId;
}

File: src/main/kotlin/org/myfintech/payment/entity/Payment.kt
/*
 * Copyright (c) 2024 MyFintech Payment Service
 * All rights reserved.
 *
 * This software is proprietary and confidential. Unauthorized copying of this file,
 * via any medium, is strictly prohibited.
 * @author : Dhanuka Ranasinghe
 * @since : Date: 11/07/2025
 */
package org.myfintech.payment.entity

import jakarta.persistence.*
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Positive
import jakarta.validation.constraints.Size
import java.math.BigDecimal
import java.time.LocalDate
import org.myfintech.payment.entity.EntityColumnConstants as Const

@Entity
@Table(name = Const.TABLE_PAYMENT)
class Payment(

    @field:NotNull
    var paymentDate: LocalDate,

    @field:NotNull
    @field:Positive
    @Column(precision = 10, scale = 2)
    var amount: BigDecimal,

    @field:NotBlank
    @field:Size(max = 50)
    var type: String,

    @Column(name = Const.FOREIGN_KEY_CONTRACT_ID, nullable = false)
    var contractId: Long,

    @Column(name = Const.FOREIGN_KEY_PAYMENT_TRACKING_ID, nullable = false)
    var trackingId: Long

) : AbstractEntity<Long>() {

    @Version
    @Column(name = Const.COMMON_VERSION, nullable = false)
    var version: Int? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = Const.FOREIGN_KEY_CONTRACT_ID, insertable = false, updatable = false)
    lateinit var contract: Contract

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = Const.FOREIGN_KEY_PAYMENT_TRACKING_ID, insertable = false, updatable = false)
    lateinit var paymentTracking: PaymentTracking


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Payment) return false
        if (!super.equals(other)) return false // Compares `id`

        // Compare explicitly included fields
        if (type != other.type) return false
        if (contractId != other.contractId) return false
        if (trackingId != other.trackingId) return false

        return true
    }

    override fun hashCode(): Int {
        // Start with superclass hash code
        var result = super.hashCode()
        // Combine with explicitly included fields
        result = 31 * result + type.hashCode()
        result = 31 * result + contractId.hashCode()
        result = 31 * result + trackingId.hashCode()
        return result
    }
}